linkerd/drain/src/lib.rs

```diff
@@ -83,184 +80,162 @@ impl Watch {
     ///
     /// The callback receives a mutable reference to the original future, and
     /// should be used to trigger any shutdown process for it.
-    pub fn watch<A, F>(self, future: A, on_drain: F) -> Watching<A, F>
+    pub async fn watch<A, F>(self, mut future: A, on_drain: F) -> A::Output
     where
-        A: Future,
+        A: Future + Unpin,
         F: FnOnce(&mut A),
     {
-        Watching {
-            future,
-            state: State::Watch(on_drain),
-            watch: self,
+        tokio::select! {
+            res = &mut future => res,
+            _ = self.rx => {
+                on_drain(&mut future);
+                (&mut future).await
+            }
```

Comment:
How do I interpret this? Is this an idiom that we're going to get intimately familiar with? If not, maybe warrants some explanation in comments...

Reply:
Oh, I think I get it. It's saying try to poll the future, and return its result, or receive the rx, and then wait on the future?